neovim: separate init.lua parts with newlines

stasjok opened the pull request:
Parts of init.lua are now separated with newlines.

### Description

Previously all parts of `init.lua` was joined without newlines.

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

ncfavier commented on modules/programs/neovim.nix:
```diff
@@ -394,26 +393,32 @@ in {
+            lib.optionalString (hasAttr "lua" cfg.generatedConfigs)
+            cfg.generatedConfigs.lua;
+          luaRcContent = builtins.concatStringsSep "\n"
+            (builtins.filter (str: str != "") [
+              vimlConfig
+              cfg.extraLuaConfig
+              luaPluginsConfig
+            ]);
```
If we're going to do this, maybe we can even add headers as comments above each section?

```nix
''
  -- Legacy vim configuration
  ${optionalString ...}

  -- Lua configuration
  ...
''
```

stasjok replied:
Usually I don't like comments with empty sections or more than two consecutive newlines (which is often a result from optional strings written this way). If I skip comments for empty strings, then it would be strange to have a comment with only one section, no? Also it is very important to not create any configs if all sections are empty.

It you think that this is a right way, I can try to do it.

teto commented:
This approach forces styling on the user. Doing nothing (as on master) grants more freedom. 
Also generating a comment block also can cause various issues down the line  depending on `type = "lua";` and the nixpkgs vimrc generator. This MR implements what used to exist for viml and I've removed in lua. 

The following snippet could instead be added to the wikii f you want to easily identify the blocks in your config:
```
  genBlockLua = title: content:
    ''
      -- ${title} {{{
      ${content}
      -- }}}
    '';

  luaPlugin = attrs: attrs // {
    type = "lua";
     config = lib.optionalString
       (attrs ? config && attrs.config != null) 
       (genBlockLua attrs.plugin.pname attrs.config)
     ;
  };
  
    defaultCompletionPlugins = with pkgs.vimPlugins; [
    (luaPlugin { plugin = nvim-cmp; })
    (luaPlugin { plugin = cmp-nvim-lsp; })
    (luaPlugin { plugin = cmp-nvim-lua; config = "-- this will generate a folded block "; })
 
```

stasjok commented:
This PR is not about style, it's about bugfix. With the following config:

```nix
{
  config = {
    programs.neovim = {
      enable = true;
      extraConfig = ''
        set sw=4
      '';
      extraLuaConfig = "vim.opt.expandtab = false";
      plugins = [{
        plugin = pkgs.emptyDirectory;
        type = "lua";
        config = ''
          vim.g.test_plugin2 = 1
        '';
      }];
    };
}
```

generated `init.lua` would be:

```lua
vim.cmd [[source /nix/store/00000000000000000000000000000000-nvim-init-home-manager.vim]]vim.opt.expandtab = falsevim.g.test_plugin2 = 1
```

I consider it an error.

It is working after `vim.cmd[[...]]`, because space is optional here in lua. But it most likely won't work without separator between `extraLuaConfig` and `generatedConfigs.lua`. User must add trailing newline, or space, or semicolon in `extraLuaConfig`. So what I did is:
1. Added two newlines after `vim.cmd[[...]]` (so it is separated with empty line, not configurable)
2. Added one newline between `extraLuaConfig` and `generatedConfigs.lua` (if user ends its `extraLuaConfig` with a trailing newline, it results in an empty line between config sections)

What I think should also be considered (I didn't do any of that so that this PR would be backward compatible):

1. Move `extraLuaConfig` first
2. Add another extra config at the end
3. Add `runtime` configuration similar to plugin's, but in top level

I think that user should be able to provide its own configuration before all plugin configurations (hence move `extraLuaConfig` to the top, because `vim.cmd [[...]]` contains plugin configs written in vimscript). Also it should be able to add config after all plugin configurations (in order to override something). To achieve all of that in my config I specified dummy empty plugins (one firsts with lua, one last with lua, and one with runtime files not linked to any plugin).

stasjok commented:
The intention of this PR was to force a newline between config section. This was addressed in #7219.
